enhancements/automatic-catalog-switching.md

```diff
@@ -0,0 +1,249 @@
+
+For example the cluster may undergo a transition from one of these releases:
+
+- Kubernetes 1.19 vs Kubernetes 1.20
+- OpenShift 4.5 vs. OpenShift 4.6
+
+When this transition occurs, a catalog image may need to be changed to represent a 
+compatible set of operators that will work properly in the new environment.
+
+For on-board OCP Operators, the change of the catalog image is automatic because it 
+is controlled by the Cluster Version Operator. Such control is missing for custom, 
+third party catalogs. This proposal describes a simple approach to 
+switching the catalog image based on platform updates.
+
+## Motivation
+
+All third party Operators have a similar requirement to automatically upgrade the Operator if 
+the underlying platform gets upgraded, to ensure users are left with a compatible and
+supported Operator install.
+
+### Goals
+
+- Enable automatic changing of the Operator catalog image as part of a platform upgrade in 
+order to avoid the problem where cluster upgrades leave Operator installations in 
+unsupported states or without a continued update path.
+- Use a simple URL templating scheme to achieve 80% of the use use case seen today
+
+### Non-Goals
+
+- Using advanced solutions
+  - a catalog of catalogs via declarative config that automatically determine which is the correct catalog
+  - new APIs that stamp out CatalogSource objects
+  - proxies at the registry side serving the catalog images
+
+## Proposal
+
+This enhancement should enable catalog image references to contain template values that can be replaced with
+actual values from the underlying cluster. These templates should allow for flexibility so that an
+image reference can be be composed in such a way that allows third parties to create image name / tag 
+in whatever way they choose. These well known template names will be replaced by a lookup 
+mechanism to discover its corresponding value. The current proposal concentrates on 
+template names that correspond to portions of a version (following semver major, minor, and patch). 
+There are other possible template values provided but version is the primary focus.
+Note that all templates must be surrounded by `{` and `}`, contain no spaces, and be all lower case.
+
+### required templates
+
+- {kube_major_version}
+  - example resolved value: 1
+- {kube_minor_version}
+  - example resolved value: 19
+- {kube_patch_version}
+  - example resolved value: 0
+- {olm_major_version}
+  - example resolved value: 0
+- {olm_minor_version}
+  - example resolved value: 18
+- {olm_patch_version}
+  - example resolved value: 1
+- {platform_architecture}
+  - example resolved value: x86_64
+
+### suggested downstream templates (only available in downstream versions of OLM)
+
+- {ocp_major_version}
+  - example resolved value: 1
+- {ocp_minor_version}
+  - example resolved value: 1
+- {ocp_patch_version}
+  - example resolved value: 1
+- {ocp_update_channels}
+  - example resolved value: fast
```

Comment:
Maybe we can omit this differentiation by making use of a generic form like `{distro_major_version}` and have OLM support this in context of the platform distribution it is running on (e.g. OKD/OCP, PKS, RKS, etc).

Reply:
We'd have to work out the "source" of the version information for each of these distributions, and the permissions needed (if any) to obtain this information. This presumes checking through a list of well known distribution types every time you need to resolve any `distro_*_version` and put an error in the status if none of them match.